File: project/app/javascript/channels/appearance_channel.ts
import consumer from "./consumer";

type Data = {
  event: "disappear" | "appear";
  user_id: number;
};

export const appearanceChannel = consumer.subscriptions.create(
  "AppearanceChannel",
  {
    connected() {
      console.log("Connected to appearance");
    },

    disconnected() {
      // Called when the subscription has been terminated by the server

      console.log("Disconnected from appearance");
    },

    received(data: Data) {
      console.log("received: ", data);
      // Called when there's incoming data on the websocket for this channel
    },
  }
);

File: project/app/channels/appearance_channel.rb
class AppearanceChannel < ApplicationCable::Channel
  def subscribed
    stream_from "appearance_channel"

    if (current_user.appearing_on == "offline")
      current_user.appear("online")
    end
  end

  def unsubscribed
    current_user.disappear
  end

  def appear(data)
    current_user.appear(data["appearing_on"])
  end

end
